controllers/persistentvolumeclaim_controller.go

```diff
@@ -0,0 +1,252 @@
+		// if schedule is not found and name is found, delete name from annotation.
+		_, nameFound := annotations[rsCronJobNameAnnotation]
+		if nameFound {
+			delete(annotations, rsCronJobNameAnnotation)
+			pvc.SetAnnotations(annotations)
+			err = r.Client.Update(ctx, pvc)
+			if err != nil {
+				logger.Error(err, "Failed to update annotation")
+
+				return ctrl.Result{}, err
+			}
+		}
+		logger.Info("Annotation not set, exiting reconcile")
+		// no schedule annotation set, just dequeue.
+		return ctrl.Result{}, nil
+	}
+
+	// create cronjob name and update in annotations.
+	rsCronJobName := fmt.Sprintf("%s-%v", pvc.Name, time.Now().Unix())
+	annotations[rsCronJobNameAnnotation] = rsCronJobName
+	pvc.SetAnnotations(annotations)
+	err = r.Client.Update(ctx, pvc)
+	if err != nil {
+		logger.Error(err, "Failed to update annotation")
+
+		return ctrl.Result{}, err
+	}
+
+	rsCronJob := constructRSCronJob(rsCronJobName, req.Namespace, schedule, pvc.Name)
+	err = ctrl.SetControllerReference(pvc, rsCronJob, r.Scheme)
+	if err != nil {
+		logger.Error(err, "Failed to set controllerReference")
+
+		return ctrl.Result{}, err
+	}
+
+	err = r.Client.Create(ctx, rsCronJob)
+	if err != nil {
+		logger.Error(err, "Failed to create reclaimSpaceCronJob")
+
+		return ctrl.Result{}, err
+	}
+
+	logger.Info("Successfully created reclaimSpaceCronJob")
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *PersistentVolumeClaimReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	if err := mgr.GetFieldIndexer().IndexField(context.Background(), &csiaddonsv1alpha1.ReclaimSpaceCronJob{}, jobOwnerKey, func(rawObj client.Object) []string {
+		// extract the owner from job object.
+		job, ok := rawObj.(*csiaddonsv1alpha1.ReclaimSpaceCronJob)
+		if !ok {
+			return nil
+		}
+		owner := metav1.GetControllerOf(job)
+		if owner == nil {
+			return nil
+		}
+		if owner.APIVersion != "v1" || owner.Kind != "PersistentVolumeClaim" {
+			return nil
+		}
+
+		return []string{owner.Name}
+	}); err != nil {
+		return err
+	}
+
+	pred := predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			if e.Object == nil {
+				return false
+			}
+			// reconcile only if schedule annotation is found.
+			_, ok := e.Object.GetAnnotations()[rsCronJobScheduleTimeAnnotation]
+			return ok
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			if e.ObjectNew == nil || e.ObjectOld == nil {
+				return false
+			}
+			// reconcile only if schedule annotation between old and new objects have changed.
+			oldSchdeule, oldOk := e.ObjectOld.GetAnnotations()[rsCronJobScheduleTimeAnnotation]
+			newSchdeule, newOk := e.ObjectNew.GetAnnotations()[rsCronJobScheduleTimeAnnotation]
+			return oldOk != newOk || oldSchdeule != newSchdeule
+		},
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&corev1.PersistentVolumeClaim{}).
+		WithEventFilter(pred).
+		Complete(r)
+}
+
+// deleteChildCronJobs lists and deletes all old child cron jobs.
```

Comment:
Describe "old child" a little more, when is it an "old child"?

Reply:
removed `old`